src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -222,67 +241,91 @@ public boolean topicExists(final String topic) throws TopicRepositoryException {
-                item.setBrokerId(partitionToBroker.get(item.getPartition()));
-            });
+    public void syncPostBatch(final String topicId, final List<BatchItem> batch, final Span parentSpan,
+                              final String eventType)
+            throws EventPublishingException {
+        final Span publishingSpan = GlobalTracer.get()
+                .buildSpan("publishing_to_kafka")
+                .asChildOf(parentSpan)
+                .start();
+        publishingSpan.setTag("event_type", eventType);
+        Tags.MESSAGE_BUS_DESTINATION.set(publishingSpan, topicId);
+        try (Scope kafkaPublishingScope = TracingService.activateSpan(publishingSpan, false)) {
+            final Producer<String, String> producer = kafkaFactory.takeProducer();
+            try {
+                final Map<String, String> partitionToBroker = producer.partitionsFor(topicId).stream().collect(
+                        Collectors.toMap(p -> String.valueOf(p.partition()), p -> String.valueOf(p.leader().id())));
+                batch.forEach(item -> {
+                    Preconditions.checkNotNull(
+                            item.getPartition(), "BatchItem partition can't be null at the moment of publishing!");
+                    item.setBrokerId(partitionToBroker.get(item.getPartition()));
+                });
 
-            int shortCircuited = 0;
-            final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
-            for (final BatchItem item : batch) {
-                item.setStep(EventPublishingStep.PUBLISHING);
-                final HystrixKafkaCircuitBreaker circuitBreaker = circuitBreakers.computeIfAbsent(
-                        item.getBrokerId(), brokerId -> new HystrixKafkaCircuitBreaker(brokerId));
-                if (circuitBreaker.allowRequest()) {
-                    sendFutures.put(item, publishItem(producer, topicId, item, circuitBreaker));
-                } else {
-                    shortCircuited++;
-                    item.updateStatusAndDetail(EventPublishingStatus.FAILED, "short circuited");
+                int shortCircuited = 0;
+                final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
+                for (final BatchItem item : batch) {
+                    item.setStep(EventPublishingStep.PUBLISHING);
+                    final HystrixKafkaCircuitBreaker circuitBreaker = circuitBreakers.computeIfAbsent(
+                            item.getBrokerId(), brokerId -> new HystrixKafkaCircuitBreaker(brokerId));
+                    if (circuitBreaker.allowRequest()) {
+                        sendFutures.put(item, publishItem(producer, topicId, item, circuitBreaker, publishingSpan));
+                    } else {
+                        shortCircuited++;
+                        item.updateStatusAndDetail(EventPublishingStatus.FAILED, "short circuited");
```

Comment:
At other places you replaced this by `handlePublishingException`. Any reason why not doing it here as well?

Reply:
fixed in b0cbd05